Add a few new reflection methods (after type system merge)

Keno opened the pull request:
This is informed by attempting to port some packages to the new type system.
I have abstracted out the functionality required by those packages into
documented/tested functions. The hope is to avoid introducing too many
implementation details to packages. These functions can also be implemented
in Compat for older julia versions, such that packages can immediate replace
their use of internal APIs with these functions.

Keno committed:
Add a few new reflection methods

This is informed by attempting to port some packages to the new type system.
I have abstracted out the functionality required by those packages into
documented/tested functions. The hope is to avoid introducing too many
implementation details to packages. These functions can also be implemented
in Compat for older julia versions, such that packages can immediate replace
their use of internal APIs with these functions.

Keno commented:
Names subject to bikeshedding of course.

JeffBezanson commented:
LGTM. These and various other type-related functions scattered around Base should probably be collected into a `TypeUtils` module eventually.

tkelman commented on base/exports.jl:
```diff
@@ -1002,6 +1002,7 @@ export
     fieldoffset,
     fieldname,
     fieldnames,
+    isabstract,
```
needs to go in stdlib doc index

tkelman commented on base/reflection.jl:
````diff
@@ -233,6 +234,55 @@ are itself and `Union{}` (but `T` itself is not `Union{}`).
+   Base.parameter_upper_bound(t::UnionAll, idx)
+
+Determine the upper bound of a type parameter in the underlying type. E.g.:
+```
````
would make a good doctest (though I don't think unexported docstrings' doctests get run by default)

ararslan commented:
Looks like you've used 3-space indents throughout the diff. For consistency it should be 4 spaces. (Indeed, AFAIK, to get the correct formatting in the resulting markdown, the signature actually has to be indented 4 spaces.)

Keno commented:
Unintentional. Sometimes editors are too smart for their own good.

Keno commented:
@JeffBezanson expressed a preference for not exporting them from base until we create the TypeUtils module. As a result, I've unexported `isabstract`. To allow packages to use this, the proposed strategy is to go ahead and create a `Compat.TypeUtils` which re-exports these and provides the Compat-ed version for old versions. In the future that can then become `const TypeUtils = Base.TypeUtils` in Compat, without requiring any code changes.

JeffBezanson commented:
Sorry, you'll have to rebase this.

Keno committed:
Ambiguity test that ignores matches due to Union{} tparams

Because of the way methods are specified, there is often spurious
ambiguities due to a type parameter being able to take the value
Union{}, (e.g. Type{T} becomes Type{Union{}}). Since detect_ambiguities
reports these, it can drown out more serious ambiguities among
non-Union{} types. Add a keyword argument to detect_ambiguities,
that ignores all ambiguities that only occur if one of the type
parameters has to be Union{}.

Keno commented:
Rebased + a new commit that allows ignoring ambiguities (in `detect_ambiguities`) due to type parameters becoming `Union{}`.

vtjnash commented on base/reflection.jl:
````diff
@@ -233,6 +234,55 @@ a concrete type that can have instances, or the empty type `Union{}`.
+```
+"""
+unparameterized_type(t::DataType) = t.name.wrapper
+unparameterized_type(t::UnionAll) = unparameterized_type(unwrap_unionall(t))
````
maybe we could try to give this a better name, although I'm not sure what. I think this is essentially `supertype`? So maybe something like `basetype`? In general, I think we've tried to avoid providing this, since it makes it tempting to assume you could re-parameterize it (even though `basetype(T){tparam}` is not generally a well-defined operation)

Keno replied:
> (even though basetype(T){tparam} is not generally a well-defined operation)

Could you elaborate on this point?

vtjnash replied:
That transform isn't meaningful for an arbitrary type `T`. It's actually not really even valid for a non-arbitrary type `T` (for example, `f(D::Type{Dict}, K, V) = D{K, V}` cannot be inferred, leading to this needed correction: https://github.com/JuliaLang/julia/pull/18457/commits/210d4ec450676c44b6011542f468c66020d82bb3#diff-d4736682004c1996e5d9b1c5c33c8c8e).

In the only cases where it is valid (e.g. you passed a type literal to `basetype`), you could just write that same type literal without the call to `basetype`. And the resulting code would also be less error-prone if someone replaced the original type with a type-alias (in addition to being shorter, by virtue of not containing no-op function calls).

Keno replied:
What's your definition of valid? That function definition you wrote down works for at least some uses.

vtjnash replied:
I mean that it's not valid to infer the output of that example function, and it's brittle to changes in the definition of `Dict`. Or consider `Vector{Int}` -- it has multiple base types, and one of the proposals for #4774 would have involved adding another. Thus you can't generally assume that you can still parameterize a type object after extracting the base type and get back something meaningful.

Keno replied:
Ok, I think I see, the point, but I'm not sure I understand why we can't infer D{K,V}.

vtjnash replied:
`D` is known to be type-equal to `Dict`, but it is not known to actually be `Dict`

ararslan commented:
@tkelman Shouldn't the base branch be master now that jb/subtype has been merged?

Keno commented:
Rebased and renamed `unparameterized_type` to `basetype` as suggested by @vtjnash

vtjnash commented:
Not to derail my bikeshedding of `unparameterized_type ` further, but maybe we should only expose `typename(T)`, since that's probably the more useful thing to extract?

Keno commented:
What would the semantics of that be?

Keno commented:
The motivating example for `basetype` here is to replace this: https://github.com/JuliaGraphics/ColorTypes.jl/blob/master/src/traits.jl#L154

vtjnash commented:
hm, that seems tricky. That code is implementing the sorts of dubious assumptions I was mentioning are usually best avoided (usually by implementing `similar`). Strictly speaking, it doesn't appear to do anything wrong, it's basically just equivalent to assuming that all subtypes of `Color` contain a field named `.parameters[1]` and that any other parameters are superfluous. This happens to work, since the subtypes implemented in Colors.jl don't contain any other parameters.

`typename` would be `typename(T::DataType) = T.name`, mainly to be used as a replacement for `datatype_name`

tkelman commented on base/reflection.jl:
```diff
@@ -233,6 +234,55 @@ a concrete type that can have instances.
+`abstract` keyword).
+"""
+function isabstract(t::ANY)
+   @_pure_meta
```
should be 4 space indent, not 3

tkelman commented on base/reflection.jl:
````diff
@@ -233,6 +234,55 @@ a concrete type that can have instances.
+```
+"""
+function parameter_upper_bound(t::UnionAll, idx)
+   @_pure_meta
````
same here

Keno committed:
Improve inference for common reflection operations

Make inference be able to infer the type constructor. This is a little
tricky, since we don't really have a good way to represent this. It's
not quite `Const(TypeVar(:T,lb,ub))`, because a) the bounds may only
be accurate up to type equality and b) TypeVar is not `isbits`, so
it's not actually egal to the value we'll have at runtime. Additionally
Type{T} already has meaning as a partially constructed type (e.g. an
unwrapped UnionAll), so using T::Type{T} runs the risk of confusing
this with an unwrapped type. Instead, introduce a new Const-like type,
`PartialTypeVar`, which carries the type var and also keeps track of
whether the bounds were egal or only typequal (we don't take advantage
of that yet, but we could in the future).

Additionally, improve the inference of `typename` and allow constant folding
of field accesses on TypeName and SimpleVectors (to be able to constant
fold T.parameters[1]).

Keno commented:
@JeffBezanson I've added some of the inference enhancements we talked about yesterday here to make these functions inferable (for ColorTypes)

tkelman commented on test/inference.jl:
```diff
@@ -491,3 +491,12 @@ tpara18457{I}(::Type{AbstractMyType18457{I}}) = I
+fComplicatedUnionAll(::Type{T}) = Type{Tuple{S,rand() >= 0.5 ? Int : Float64}} where S <: T
+@test Base.parameter_upper_bound(fComplicatedUnionAll(Real), 1) == Real
+let x = fComplicatedUnionAll(Real).parameters[1].parameters[2]
+  @test x === Int || x === Float64
```
4 space indent

vtjnash commented on test/inference.jl:
```diff
@@ -491,3 +491,12 @@ tpara18457{I}(::Type{AbstractMyType18457{I}}) = I
+
+fComplicatedUnionAll(::Type{T}) = Type{Tuple{S,rand() >= 0.5 ? Int : Float64}} where S <: T
+@test Base.parameter_upper_bound(fComplicatedUnionAll(Real), 1) == Real
+let x = fComplicatedUnionAll(Real).parameters[1].parameters[2]
```
This test should be for `parameters[1] == Tuple{Real, Int} or Tuple{Real, Float64}`, since it is not true that `typeof(parameters[1]) === DataType`. (it it permitted to instead be `Tuple{T, Int} where T <: Real`)

Keno replied:
Yes this test is wrong. Fixing now

vtjnash commented on base/inference.jl:
```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
+            val = isa(t, Const) ? t.val : t.parameters[1]
+            try
+                return Const(typename(val))
+            catch
```
don't use try/catch when you can avoid it (weren't you also the one complaining that it makes debugging harder?)

Keno replied:
So you're suggesting repeating the logic here that does typename?

vtjnash replied:
In a helper function, but yes. You have to repeat most of it to figure out why it failed anyways (TypeVars, Unions of TypeVars, Vararg uncertainty, etc.)

vtjnash commented on base/inference.jl:
```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
+    elseif length(argtypes) == 2 && istopfunction(tm, f, :typename)
+        t = argtypes[2]
+        if isa(t, Const) || isType(t)
+            val = isa(t, Const) ? t.val : t.parameters[1]
```
this might encounter `Vararg` or `TypeVar`, and will need to handle those cases also

vtjnash commented on base/inference.jl:
```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
 
     if istopfunction(tm, f, :promote_type) || istopfunction(tm, f, :typejoin)
         return Type
+    elseif length(argtypes) == 2 && istopfunction(tm, f, :typename)
```
we should consider trying to handle this by marking the function `@pure` instead of doing a partial reimplementation here and trying to deal with all of the same edge cases

Keno replied:
It's not just pure though, it also maps the type equality equivalence class to one value.

vtjnash commented on base/inference.jl:
```diff
@@ -1411,7 +1469,15 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
+            if isa(argtypes[2], Const) && isa(argtypes[3], Const)
+                svecval = argtypes[2].val
+                idx = argtypes[3].val
+                if isa(idx, Int) && 1 <= idx <= length(svecval)
```
`&& isdefined(svecval, idx)`

vtjnash replied:
although, actually, why isn't this being handled by the getfield_tfunc tfunc?

Keno replied:
getfield on an svec is a different function (try it). I was confused by that too at first

vtjnash replied:
Ah, true. Sight, the really annoying part here is going to be that for every extra case here, you need to duplicate it and add a special case for inlineable too since they need to model each other (for example, https://github.com/JuliaLang/julia/pull/20006/files#diff-c0d983631c937ee5acfd0137cd7332deR1515)

Keno replied:
Is is actually possible to have an undefined element in an svec?

vtjnash replied:
sadly, yes

Keno replied:
Ok, but isdefined doesn't check for that? Unless you mean that the .length field could be undefined. E.g.

```
julia> isdefined(Core.svec(1,2),2)
false
```

vtjnash commented on base/inference.jl:
```diff
@@ -1386,19 +1401,62 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
+                return Any
+            end
+            !isa(tv, TypeVar) && return Any
+            ret = canconst ? abstract_eval_constant(UnionAll(tv, body)) :
```
`UnionAll` could throw if `body` isn't a Type

vtjnash commented on base/inference.jl:
```diff
@@ -671,7 +679,8 @@ function getfield_tfunc(s00::ANY, name)
                 return abstract_eval_global(sv, nv)
             end
-            if (isa(sv, DataType) || isimmutable(sv)) && isdefined(sv, nv)
+            if (isa(sv, DataType) || isa(sv, SimpleVector) || isa(sv, TypeName)
```
the `cache` and `linearcache` fields of TypeName are not constant

although, I guess neither is the `uid` field of DataType

JeffBezanson commented on base/inference.jl:
```diff
@@ -63,6 +63,13 @@ immutable Const
+immutable PartialTypeVar
+    tv::TypeVar
+    lb_certain::Bool
+    ub_certain::Bool
```
I would define this as

```
immutable PartialTypeVar
    name
    lb
    ub
end
```

where the components are simply the inferred types of the arguments to `TypeVar`. You know they're certain if they're `Const`. Did you come across any reasons this approach wouldn't work?

vtjnash replied:
If I'm not mistaken, Keno is using the identity of the TypeVar to infer the equivalence class of the result of `UnionAll(tv, apply_type(Type, tv)))`

Keno replied:
Yes, what @vtjnash said. You need to track identity through apply_type such that the UnionAll constructor works.

JeffBezanson commented on base/inference.jl:
```diff
@@ -1386,19 +1401,62 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
+            elseif isa(argtypes[2], PartialTypeVar)
+                ptv = argtypes[2]
+                tv = ptv.tv
+                canconst = false
```
Does this depend on `lb_certain` and `ub_certain`?

Keno replied:
No. Two invocations of this functions do not create egal types (because it allocates a new one every time). I though we could do something with that at first, but I had to take that out because of that.

Keno commented:
I've addressed some of the review here. @JeffBezanson @vtjnash, I'd appreciate if you could finish this up, since I'll be traveling the rest of this week and this is required for ColorTypes.

Keno committed:
Don't constant-fold non-constant fields

for DataType/TypeName

tkelman commented on base/inference.jl:
```diff
@@ -1617,6 +1702,12 @@ function abstract_call(f::ANY, fargs::Union{Tuple{},Vector{Any}}, argtypes::Vect
+        t = argtypes[2]
+        if isa(t, Const) || isType(t)
+            return typename_static(t)
+         end
```
indent is a bit off

Keno replied:
So it is. Sorry, I'm used to sublime doing this all automatically for me, but I've been trying to use Atom more. Will fix

Keno commented:
Review comments addressed.

JeffBezanson commented on base/inference.jl:
```diff
@@ -82,6 +82,13 @@ type Conditional
+    tv::TypeVar
+    lb_certain::Bool
+    ub_certain::Bool
+    PartialTypeVar(tv::TypeVar, lb_certain::ANY, ub_certain::ANY) = new(tv, lb_certain, ub_certain)
```
Why use `ANY` here if the arguments have to be Bool?

Keno replied:
No reason. Will change.

JeffBezanson commented on base/inference.jl:
```diff
@@ -1476,6 +1512,25 @@ function Pair_name()
+function _typename(a::Union)
+    ta = _typename(a.a)
+    tb = _typename(a.b)
+    ta == tb ? tb : (ta === Any || tb == Any) ? Any : Union{}
```
Should this use `ta === tb`? (also `tb === Any` for consistency)

Keno replied:
No, I don't this it should since since `Const(T.name) != Const(T.name)`. `tb === Any is fine though`.

JeffBezanson commented on test/inference.jl:
```diff
@@ -588,3 +588,14 @@ f11015(a::AT11015) = g11015(Base.fieldtype(typeof(a), :f), true)
+
+fComplicatedUnionAll{T}(::Type{T}) = Type{Tuple{S,rand() >= 0.5 ? Int : Float64}} where S <: T
+let pub = Base.parameter_upper_bound, x = fComplicatedUnionAll(Real)
+    @test pub(pub(x, 1), 1) == Real
```
Do these belong in the reflection tests instead?

Keno replied:
I put them here, because at some point I had inference mess up and constant fold this to the wrong answer.

Keno committed:
Merge branch 'master' into kf/newreflection

tkelman commented on base/inference.jl:
```diff
@@ -688,10 +697,25 @@ function limit_type_depth(t::ANY, d::Int, cov::Bool=true, var::Union{Void,TypeVa
+const TypeName_wrapper_fieldindex = fieldindex(TypeName, :wrapper)
+
+function const_datatype_getfield_tfunc(sv, fld)
+  if (fld == DataType_name_fieldindex ||
```
wrong indent, again
